narwhal/primary/src/core.rs

```diff
@@ -139,11 +139,30 @@ impl Core {
                 cancel_handlers: HashMap::with_capacity(2 * gc_depth as usize),
                 metrics,
             }
+            .recover()
+            .await
             .run()
             .await;
         })
     }
 
+    #[instrument(level = "info", skip_all)]
+    pub async fn recover(mut self) -> Self {
+        info!("Starting certificate recovery. Message processing will begin after completion.");
+
+        let last_round_certificates = self
+            .certificate_store
+            .last_round()
+            .expect("Failed recovering certificates in primary core");
```

Comment:
How do we ensure that the proposer won't create two conflicting headers? Through your other PR?

Reply:
If so, should we merge them? If anything goes wrong, it will be easier to traceback bugs